failed to drain node with multiple nics

<!--
You don't need to remove this comment section, it's invisible on the issues page.

## General remarks

* Attention, please fill out this issues form using English!
* 注意！GitHub Issue 请使用英文
-->

**Describe the Bug**
node1: ip1, ip2,ip3
kcctl join -agent bj:ip2
after joined, the node1 is added with ip3

after that, when kcctl drain -agent ip3, it failed to do it.

the actual info is shown as the following:

```
root@compute1:~# kcctl get node
+--------------------------------------+----------+---------+---------------+-------------+-----+---------+
|                  ID                  | HOSTNAME | REGION  |      IP       |   OS/ARCH   | CPU |   MEM   |
+--------------------------------------+----------+---------+---------------+-------------+-----+---------+
| 8dbf9601-cd9a-44f4-8033-3c42f6eca7c0 | compute4 | bj      | 192.168.52.14 | linux/amd64 |   8 | 16008Mi |
| ae337ab9-4c6b-4902-9cc4-b9d5b311376c | compute1 | default | 192.168.51.11 | linux/amd64 |   8 | 16008Mi |
+--------------------------------------+----------+---------+---------------+-------------+-----+---------+
root@compute1:~# kcctl drain  --agent compute4
[2022-09-07T08:42:45Z][ERROR] agent compute4 is not in agent nodes
root@compute1:~# kcctl drain  --agent 192.168.52.14
[2022-09-07T08:42:58Z][ERROR] agent 192.168.52.14 is not in agent nodes
root@compute1:~# kcctl drain  --agent 192.168.51.14
drain agent node failed: the node could not be draind. reason: 192.168.51.14 does not exist


```

**Versions Used**
Kubernetes:
Kubeclipper:

```
root@compute1:~# kcctl version -o json
kcctl version:
{
	"major": "1",
	"minor": "1",
	"gitVersion": "v1.1.0",
	"gitCommit": "e5dab3cc8a5e200797c3c2b2c151b8356dbf436b",
	"gitTreeState": "clean",
	"buildDate": "2022-08-13T06:43:05Z",
	"goVersion": "go1.18.5",
	"compiler": "gc",
	"platform": "linux/amd64"
}
kubeclipper-server version:
{
	"major": "1",
	"minor": "1",
	"gitVersion": "v1.1.0",
	"gitCommit": "e5dab3cc8a5e200797c3c2b2c151b8356dbf436b",
	"gitTreeState": "clean",
	"buildDate": "2022-08-13T06:31:44Z",
	"goVersion": "go1.18.5",
	"compiler": "gc",
	"platform": "linux/amd64"
}

```


**Environment**
How many nodes and their hardware configuration:

For example: CentOS 7.5 / 3 masters:  8cpu/8g; 3 nodes: 8cpu/16g
(and other info are welcomed to help us debugging)

**How To Reproduce**
Steps to reproduce the behavior:
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

--- a/cmd/kcctl/app/options/options.go
+++ b/cmd/kcctl/app/options/options.go
@@ -185,12 +185,29 @@ func (a Agents) ListIP() []string {
 	}
 	return list
 }
+
 func (a Agents) Exists(ip string) bool {
 	_, ok := a[ip]
 	return ok
 }
-func (a Agents) Delete(ip string) {
+
+func (a Agents) ExistsByID(id string) bool {
+	return a.idToIP(id) != ""
+}
+
+func (a Agents) Delete(id string) {
+	ip := a.idToIP(id)
 	delete(a, ip)
+}
+
+func (a Agents) idToIP(id string) string {
+	var ip string
+	for k, v := range a {
+		if v.AgentID == id {
+			ip = k
+		}
+	}
+	return ip
 }
 
 func (a Agents) Add(ip string, metadata Metadata) {
@@ -199,6 +216,7 @@ func (a Agents) Add(ip string, metadata Metadata) {
 
 // Metadata user custom node info,region will use by filter,use label,others use annotation.
 type Metadata struct {
+	AgentID string `json:"agentID" yaml:"agentID,omitempty"`
 	Region  string `json:"region" yaml:"region,omitempty"`
 	FloatIP string `json:"floatIP" yaml:"floatIP,omitempty"`
 
@@ -369,12 +387,11 @@ func (c *DeployConfig) Write() error {
 	if c.Config == "" {
 		path = DefaultDeployConfigPath
 	}
-
 	b, err := yaml.Marshal(c)
 	if err != nil {
 		return fmt.Errorf("dump config failed due to %s", err.Error())
 	}
-	if err := utils.WriteToFile(path, b); err != nil {
+	if err = utils.WriteToFile(path, b); err != nil {
 		return fmt.Errorf("dump config to %s failed due to %s", path, err.Error())
 	}
 	return nil
@@ -495,7 +512,7 @@ func (c *DeployConfig) GetKcServerConfigTemplateContent(ip string) (string, erro
 	return buffer.String(), nil
 }
 
-func (c *DeployConfig) GetKcAgentConfigTemplateContent(metadata Metadata, agentID string) (string, error) {
+func (c *DeployConfig) GetKcAgentConfigTemplateContent(metadata Metadata) (string, error) {
 	tmpl, err := template.New("text").Parse(config.KcAgentConfigTmpl)
 	if err != nil {
 		return "", fmt.Errorf("template parse failed: %s", err.Error())
@@ -506,7 +523,7 @@ func (c *DeployConfig) GetKcAgentConfigTemplateContent(metadata Metadata, agentI
 	}
 
 	var data = make(map[string]interface{})
-	data["AgentID"] = agentID
+	data["AgentID"] = metadata.AgentID
 	data["Region"] = metadata.Region
 	data["FloatIP"] = metadata.FloatIP
 	data["IPDetect"] = c.IPDetect

--- a/pkg/apis/core/v1/handler.go
+++ b/pkg/apis/core/v1/handler.go
@@ -33,6 +33,15 @@ import (
 	"github.com/emicklei/go-restful"
 	"github.com/google/uuid"
 	"github.com/gorilla/websocket"
+	"github.com/robfig/cron/v3"
+	"go.uber.org/zap"
+	apimachineryErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+	r "k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/remotecommand"
+
 	"github.com/kubeclipper/kubeclipper/pkg/client/clientrest"
 	"github.com/kubeclipper/kubeclipper/pkg/clustermanage"
 	"github.com/kubeclipper/kubeclipper/pkg/component"
@@ -58,14 +67,6 @@ import (
 	"github.com/kubeclipper/kubeclipper/pkg/utils/netutil"
 	"github.com/kubeclipper/kubeclipper/pkg/utils/sshutils"
 	"github.com/kubeclipper/kubeclipper/pkg/utils/strutil"
-	"github.com/robfig/cron/v3"
-	"go.uber.org/zap"
-	apimachineryErrors "k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/util/json"
-	r "k8s.io/apimachinery/pkg/util/rand"
-	"k8s.io/apimachinery/pkg/util/sets"
-	"k8s.io/client-go/tools/remotecommand"
 )
 
 type handler struct {

--- a/pkg/apis/core/v1/utils.go
+++ b/pkg/apis/core/v1/utils.go
@@ -25,6 +25,7 @@ import (
 	"path/filepath"
 
 	"github.com/google/uuid"
+
 	"github.com/kubeclipper/kubeclipper/pkg/models/cluster"
 
 	"github.com/kubeclipper/kubeclipper/pkg/query"

--- a/pkg/cli/deploy/deploy.go
+++ b/pkg/cli/deploy/deploy.go
@@ -56,8 +56,6 @@ import (
 	"github.com/sethvargo/go-password/password"
 	"gopkg.in/yaml.v2"
 
-	"github.com/kubeclipper/kubeclipper/pkg/utils/cmdutil"
-
 	"github.com/kubeclipper/kubeclipper/pkg/cli/join"
 
 	"github.com/kubeclipper/kubeclipper/pkg/cli/sudo"
@@ -440,6 +438,8 @@ func (d *DeployOptions) RunDeploy() error {
 	d.deployKcConsole()
 	logger.Infof("------ Delete intermediate files ------")
 	d.removeTempFile()
+	logger.Infof("------ Dump configs ------")
+	d.dumpConfig()
 	logger.Infof("------ Upload configs ------")
 	d.uploadConfig()
 	fmt.Printf("\033[1;40;36m%s\033[0m\n", options.Contact)
@@ -697,9 +697,11 @@ func (d *DeployOptions) deployKcConsole() {
 }
 
 func (d *DeployOptions) deployKcAgent() {
-	for agent, metadata := range d.deployConfig.Agents {
-		agentID := uuid.New().String()
-		agentConfig, err := d.deployConfig.GetKcAgentConfigTemplateContent(metadata, agentID)
+	for agent := range d.deployConfig.Agents {
+		metadata := d.deployConfig.Agents[agent]
+		metadata.AgentID = uuid.New().String()
+		d.deployConfig.Agents[agent] = metadata
+		agentConfig, err := d.deployConfig.GetKcAgentConfigTemplateContent(metadata)
 		if err != nil {
 			logger.Fatal(err)
 		}
@@ -734,16 +736,8 @@ func (d *DeployOptions) removeTempFile() {
 }
 
 func (d *DeployOptions) dumpConfig() {
-	_, err := cmdutil.RunCmd(false, "mkdir", "-p", filepath.Dir(options.DefaultDeployConfigPath))
-	if err != nil {
-		logger.Fatalf("dump config failed due to %s", err.Error())
-	}
-	b, err := yaml.Marshal(d.deployConfig)
-	if err != nil {
-		logger.Fatalf("dump config failed due to %s", err.Error())
-	}
-	if err = utils.WriteToFile(options.DefaultDeployConfigPath, b); err != nil {
-		logger.Fatalf("dump config to %s failed due to %s", options.DefaultDeployConfigPath, err.Error())
+	if err := d.deployConfig.Write(); err != nil {
+		logger.Fatal(err)
 	}
 	logger.V(2).Infof("dump config to %s", options.DefaultDeployConfigPath)
 }

--- a/pkg/cli/deploy/deploy_test.go
+++ b/pkg/cli/deploy/deploy_test.go
@@ -67,8 +67,8 @@ func TestDeployOptions_getKcAgentConfigTemplateContent(t *testing.T) {
 		FloatIP: "1.1.1.1",
 	}
 	for range d.deployConfig.ServerIPs {
-		agentID := uuid.New().String()
-		t.Log(d.deployConfig.GetKcAgentConfigTemplateContent(metadata, agentID))
+		metadata.AgentID = uuid.New().String()
+		t.Log(d.deployConfig.GetKcAgentConfigTemplateContent(metadata))
 	}
 }
 

--- a/pkg/cli/drain/drain.go
+++ b/pkg/cli/drain/drain.go
@@ -45,17 +45,17 @@ import (
 const (
 	longDescription = `
   Drain the Kubeclipper server or agent node from the cluster.
-
+ 
   Now only support drain kc-agent node, so the --agent parameter must be valid.`
 	drainExample = `
   # Drain kc-agent from kubeclipper cluster use default deploy-config(~/.kc/deploy-config.yaml) and config(~/.kc/config).
-  kcctl drain --agent 192.168.10.19
+  kcctl drain --agent d55e10ec-4e7e-4ce9-9ce7-eb491ddc7bfa
 
   # Drain kc-agent from kubeclipper cluster specify deploy config and config.
-  kcctl drain --agent 192.168.10.19 --deploy-config /root/.kc/deploy-config.yaml --config /root/.kc/config
+  kcctl drain --agent d55e10ec-4e7e-4ce9-9ce7-eb491ddc7bfa --deploy-config /root/.kc/deploy-config.yaml --config /root/.kc/config
 
   # Force drain kc-agent which is in used from kubeclipper cluster
-  kcctl drain  --force --agent=192.168.10.123
+  kcctl drain  --force --agent=d55e10ec-4e7e-4ce9-9ce7-eb491ddc7bfa
 
   Please read 'kcctl drain -h' get more drain flags.`
 )
@@ -81,7 +81,7 @@ func NewDrainOptions(streams options.IOStreams) *DrainOptions {
 func NewCmdDrain(streams options.IOStreams) *cobra.Command {
 	o := NewDrainOptions(streams)
 	cmd := &cobra.Command{
-		Use:                   "drain (--agent <agentIps>) [flags]",
+		Use:                   "drain (--agent <agentIDs>) [flags]",
 		DisableFlagsInUseLine: true,
 		Short:                 "Drain kubeclipper agent node",
 		Long:                  longDescription,
@@ -97,7 +97,7 @@ func NewCmdDrain(streams options.IOStreams) *cobra.Command {
 	}
 
 	o.cliOpts.AddFlags(cmd.Flags())
-	cmd.Flags().StringSliceVar(&o.agents, "agent", o.agents, "drain agent node ip.")
+	cmd.Flags().StringSliceVar(&o.agents, "agent", o.agents, "drain agent node ID.")
 	cmd.Flags().StringVar(&o.deployConfig.Config, "deploy-config", options.DefaultDeployConfigPath, "kcctl deploy config path")
 	cmd.Flags().BoolVarP(&o.force, "force", "F", o.force, "force delete in used node.")
 
@@ -160,14 +160,24 @@ func (c *DrainOptions) ValidateArgs() error {
 	if len(c.agents) == 0 {
 		return errors.New("--agent is required")
 	}
+
+	for _, agent := range c.agents {
+		if !isAgentID(agent) {
+			return fmt.Errorf("--agent need a node id, %s is invalid", agent)
+		}
+	}
 	return nil
+}
+
+func isAgentID(id string) bool {
+	return len(id) == 36
 }
 
 func (c *DrainOptions) preCheck() bool {
 	c.agents = sets.NewString(c.agents...).List()
 
 	for _, agent := range c.agents {
-		if !c.deployConfig.Agents.Exists(agent) {
+		if !c.deployConfig.Agents.ExistsByID(agent) {
 			logger.Errorf("agent %s is not in agent nodes", agent)
 			return false
 		}
@@ -211,24 +221,20 @@ func (c *DrainOptions) runDrainAgentNode() error {
 	return nil
 }
 
-func (c *DrainOptions) checkAgentNode(ip string) (*v1.Node, error) {
-	q := query.New()
-	nodeList, err := c.client.ListNodes(context.TODO(), kc.Queries(*q))
+func (c *DrainOptions) checkAgentNode(id string) (*v1.Node, error) {
+	nodeList, err := c.client.DescribeNode(context.TODO(), id)
 	if err != nil {
 		return nil, err
 	}
-	for _, node := range nodeList.Items {
-		// check the node already exists in database
-		if node.Status.Ipv4DefaultIP == ip {
-			// check the node already is used
-			if _, ok := node.Labels[common.LabelNodeRole]; !ok || c.force {
-				return &node, nil
-			}
-			return nil, fmt.Errorf("the node could not be draind. reason: %s is used by the cluster", ip)
-		}
+	if len(nodeList.Items) == 0 {
+		return nil, fmt.Errorf("the node could not be draind. reason: %s does not exist", id)
 	}
-
-	return nil, fmt.Errorf("the node could not be draind. reason: %s does not exist", ip)
+	node := nodeList.Items[0]
+	clusterName, ok := node.Labels[common.LabelClusterName]
+	if !ok || c.force {
+		return &node, nil
+	}
+	return nil, fmt.Errorf("the node could not be draind. reason: %s is used by the cluster %s", id, clusterName)
 }
 
 func (c *DrainOptions) agentFilesAndData(node *v1.Node) error {
@@ -252,7 +258,7 @@ func (c *DrainOptions) agentFilesAndData(node *v1.Node) error {
 	}
 
 	// 	3.rewrite deploy config
-	c.deployConfig.Agents.Delete(node.Status.Ipv4DefaultIP)
+	c.deployConfig.Agents.Delete(node.Name)
 	err = c.deployConfig.Write()
 	if err != nil {
 		return errors.WithMessage(err, "rewrite deploy config")

--- a/pkg/cli/join/join.go
+++ b/pkg/cli/join/join.go
@@ -192,7 +192,10 @@ func (c *JoinOptions) RunJoinNode() error {
 
 func (c *JoinOptions) runJoinAgentNode() error {
 	var err error
-	for ip, metadata := range c.parseAgent {
+	for ip := range c.parseAgent {
+		metadata := c.parseAgent[ip]
+		metadata.AgentID = uuid.New().String()
+		c.parseAgent[ip] = metadata
 		if err = c.agentNodeFiles(ip, metadata); err != nil {
 			return err
 		}
@@ -295,7 +298,7 @@ func (c *JoinOptions) getKcAgentConfigTemplateContent(metadata options.Metadata)
 	data["Region"] = metadata.Region
 	data["FloatIP"] = metadata.FloatIP
 	data["IPDetect"] = c.deployConfig.IPDetect
-	data["AgentID"] = uuid.New().String()
+	data["AgentID"] = metadata.AgentID
 	data["StaticServerAddress"] = fmt.Sprintf("http://%s:%d", c.deployConfig.ServerIPs[0], c.deployConfig.StaticServerPort)
 	if c.deployConfig.Debug {
 		data["LogLevel"] = "debug"

--- a/pkg/cli/utils/file.go
+++ b/pkg/cli/utils/file.go
@@ -20,6 +20,7 @@ package utils
 
 import (
 	"os"
+	"path/filepath"
 )
 
 func FileExist(path string) bool {
@@ -27,8 +28,12 @@ func FileExist(path string) bool {
 	return !os.IsNotExist(err)
 }
 
-func WriteToFile(filePath string, data []byte) error {
-	f, err := os.Create(filePath)
+func WriteToFile(path string, data []byte) error {
+	err := os.MkdirAll(filepath.Dir(path), 0777)
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(path)
 	if err != nil {
 		return err
 	}

--- a/pkg/clustermanage/kubeadm/provider.go
+++ b/pkg/clustermanage/kubeadm/provider.go
@@ -406,7 +406,7 @@ func (r Kubeadm) deployKCAgent(ctx context.Context, node *v1.WorkerNode, region 
 	}
 
 	// 2. generate kubeclipper-agent.yaml、systemd conf,then start kc-agent
-	agentConfig, err := deployConfig.GetKcAgentConfigTemplateContent(options.Metadata{Region: region}, node.ID)
+	agentConfig, err := deployConfig.GetKcAgentConfigTemplateContent(options.Metadata{Region: region, AgentID: node.ID})
 	if err != nil {
 		return errors.WithMessage(err, "GetKcAgentConfigTemplateContent")
 	}
@@ -461,7 +461,7 @@ func (r Kubeadm) drainAgent(nodeIP, agentID string, ssh *sshutils.SSH) error {
 		return errors.WithMessage(err, "getDeployConfig")
 	}
 
-	if !deployConfig.Agents.Exists(nodeIP) {
+	if deployConfig.Agents.ExistsByID(nodeIP) {
 		err = r.updateDeployConfigAgents(nodeIP, nil, "del")
 		if err != nil {
 			logger.Errorf("add agent ip to deploy config failed: %v", err)

--- a/pkg/controller/cluster_status.go
+++ b/pkg/controller/cluster_status.go
@@ -25,9 +25,10 @@ import (
 	"strings"
 	"time"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
 	"github.com/kubeclipper/kubeclipper/pkg/clustermanage"
 	"github.com/kubeclipper/kubeclipper/pkg/scheme/common"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/kubeclipper/kubeclipper/pkg/scheme/core/v1/k8s"
 

--- a/pkg/controller/clustercontroller/controller.go
+++ b/pkg/controller/clustercontroller/controller.go
@@ -24,10 +24,11 @@ import (
 	"strings"
 	"time"
 
-	"github.com/kubeclipper/kubeclipper/pkg/clustermanage"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/kubeclipper/kubeclipper/pkg/clustermanage"
 
 	"github.com/kubeclipper/kubeclipper/pkg/controller-runtime/reconcile"
 

